src/gps.c

```diff
@@ -0,0 +1,279 @@
+        {
+          // Server is not responding ...
+          if (gps_send (&gpsd_conn, GPS_CONFIG) == -1)
+          {
+            WARNING ("gps plugin: gpsd seems to be down, reconnecting");
+            gps_close (&gpsd_conn);
+            break;
+          }
+          // Server is responding ...
+          else
+          {
+            err_count = 0;
+          }
+        }
+
+        continue;
+      }
+
+      pthread_mutex_lock (&data_lock);
+
+      // Number of sats in view:
+      data.sats_used = (gauge_t) gpsd_conn.satellites_used;
+      data.sats_visible = (gauge_t) gpsd_conn.satellites_visible;
+
+      // dilution of precision:
+      data.vdop = NAN; data.hdop = NAN;
+      if (data.sats_used > 0)
+      {
+        data.hdop = gpsd_conn.dop.hdop;
+        data.vdop = gpsd_conn.dop.vdop;
+      }
+
+
+      DEBUG ("gps plugin: %.0f sats used (of %.0f visible), hdop = %.3f, vdop = %.3f",
+             data.sats_used, data.sats_visible, data.hdop, data.vdop);
+
+      pthread_mutex_unlock (&data_lock);
+    }
+  }
+
+  gps_stream (&gpsd_conn, WATCH_DISABLE, /* data = */ NULL);
+  gps_close (&gpsd_conn);
+
+  pthread_exit ((void *) 0);
+}
+
+/**
+ * Submit a piece of the data.
+ */
+static void cgps_submit (const char *type, gauge_t value, const char *type_instance)
+{
+  value_t values[1];
+  value_list_t vl = VALUE_LIST_INIT;
+
+  values[0].gauge = value;
+
+  vl.values = values;
+  vl.values_len = 1;
+  sstrncpy (vl.host, hostname_g, sizeof (vl.host));
+  sstrncpy (vl.plugin, "gps", sizeof (vl.plugin));
+  sstrncpy (vl.type, type, sizeof (vl.type));
+  sstrncpy (vl.type_instance, type_instance, sizeof (vl.type_instance));
+
+  plugin_dispatch_values (&vl);
+}
+
+/**
+ * Read the data and submit by piece.
+ */
+static int cgps_read ()
+{
+  cgps_data_t data_copy;
+
+  pthread_mutex_lock (&data_lock);
+  data_copy = data;
+  pthread_mutex_unlock (&data_lock);
+
+  cgps_submit ("dilution_of_precision", data_copy.hdop, "horizontal");
+  cgps_submit ("dilution_of_precision", data_copy.vdop, "vertical");
+  cgps_submit ("satellites", data_copy.sats_used, "used");
+  cgps_submit ("satellites", data_copy.sats_visible, "visible");
+
+  return (0);
+}
+
+/**
+ * Read configuration.
+ */
+static int cgps_config (oconfig_item_t *ci)
+{
+  int i;
+
+  for (i = 0; i < ci->children_num; i++)
+  {
+    oconfig_item_t *child = ci->children + i;
+
+    if (strcasecmp ("Host", child->key) == 0)
+      cf_util_get_string (child, &config.host);
+    else if (strcasecmp ("Port", child->key) == 0)
+      cf_util_get_service (child, &config.port);
+    else if (strcasecmp ("Timeout", child->key) == 0)
+      cf_util_get_cdtime (child, &config.timeout);
+    else if (strcasecmp ("Pause", child->key) == 0)
+      cf_util_get_cdtime (child, &config.pause);
+    else
+      WARNING ("gps plugin: Ignoring unknown config option \"%s\".", child->key);
+  }
+
+  return 0;
+}
+
+/**
+ * Init.
+ */
+static int cgps_init (void)
```

Comment:
The *init* callback may be called multiple times. I'd propose to use the `connector_running` boolean mentioned below to protect against multiple calls of this callback:

```c
if (connector_running)
  return 0;
```